java-frontend/src/main/java/org/sonar/java/regex/JavaCharacterParser.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012-2020 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.regex;
+
+import org.sonar.java.regex.ast.IndexRange;
+import org.sonar.java.regex.ast.JavaCharacter;
+import org.sonar.java.regex.ast.RegexSource;
+
+/**
+ * Parse the contents of string literals and provide the individual characters of the string after processing escape
+ * sequences
+ */
+public class JavaCharacterParser {
+
+  public static final int EOF = -1;
+
+  private final RegexSource source;
+
+  private final JavaUnicodeEscapeParser unicodeProcessedCharacters;
+
+  private JavaCharacter current;
```

Comment:
I would annotate this field with `@Nullable` and explicitly document when it will be null (end of source?)